pkg/ratelimit/resource.go

```diff
@@ -44,18 +47,51 @@ type resource struct {
 func (r *resource) Key() string       { return r.key }
 func (r *resource) Classes() []string { return r.classes }
 
+const unauthenticated = "unauthenticated"
+
+func authTokenID(ctx context.Context) string {
+	if authValue := rpcmetadata.FromIncomingContext(ctx).AuthValue; authValue != "" {
+		_, id, _, err := auth.SplitToken(authValue)
+		if err != nil {
+			return unauthenticated
+		}
+		return id
+	}
+	return unauthenticated
+}
+
+var pathTemplate = func(r *http.Request) (string, bool) {
+	route := mux.CurrentRoute(r)
+	if route == nil {
+		return "", false
+	}
+	pathTemplate, err := route.GetPathTemplate()
+	if err != nil {
+		return "", false
+	}
+	return pathTemplate, true
+}
+
 // httpRequestResource represents an HTTP request. Avoid using directly, use HTTPMiddleware instead.
 func httpRequestResource(r *http.Request, class string) Resource {
+	specificClasses := make([]string, 0, 3)
+	if template, ok := pathTemplate(r); ok {
+		specificClasses = append(specificClasses, fmt.Sprintf("%s:%s", class, template))
+	}
+	authKey := fmt.Sprintf("ip:%s", httpRemoteIP(r))
+	if authTokenID := authTokenID(r.Context()); authTokenID != unauthenticated {
+		authKey = fmt.Sprintf("token:%s", authTokenID)
+	}
 	return &resource{
-		key:     fmt.Sprintf("%s:ip:%s:url:%s", class, httpRemoteIP(r), r.URL.Path),
-		classes: []string{class, "http"},
+		key:     fmt.Sprintf("%s:%s:url:%s", class, authKey, r.URL.Path),
+		classes: append(specificClasses, class, "http"),
```

Comment:
So in the case where the template can be extracted, this becomes `"class:template:class:http"`?

Reply:
A single resource can be part of multiple classes. The idea is to have a form of hierarchy:
- `http` is common to all HTTP endpoints. You would have the highest generic rate limit here.
- `http:gcs` is the common class for all endpoints of the GCS. You would have an intermediate rate limit here.
- `http:gcs:/some-path-here` is the specific class of an endpoint. You would have the lowest rate limit here, if needed.